agent/xdsv2/rbac_resources.go

```diff
@@ -169,6 +174,196 @@ func makeRBACPolicy(p *pbproxystate.Permission) *envoy_rbac_v3.Policy {
 	}
 }
 
+func makeL7RBACPolicies(l7Permissions []*pbproxystate.Permission) map[string]*envoy_rbac_v3.Policy {
+	// sort permissions into those with L7-specific features and those without, to match labeling and behavior
+	// conventions in V1: https://github.com/hashicorp/consul/blob/4e451f23584473a7eaf7f123145ca85e0a31783a/agent/xds/rbac.go#L647
+	// this is a somewhat unfortunate carry-over needed for testing v1 vs v2 final config
+	// and this will break with v1 intentions when multiple L4 permissions are used
+	var l4Perms []*pbproxystate.Permission
+	var l7Perms []*pbproxystate.Permission
+	for _, p := range l7Permissions {
+		if len(p.DestinationRules) > 0 {
+			l7Perms = append(l7Perms, p)
+		} else {
+			l4Perms = append(l4Perms, p)
+		}
+	}
+	l4PolicyLabel := func(i int) string {
+		if len(l4Perms) == 1 {
+			return baseL4PermissionKey
+		}
+		return fmt.Sprintf("%s-%d", baseL4PermissionKey, i)
+	}
+
+	policies := make(map[string]*envoy_rbac_v3.Policy, len(l7Permissions))
+
+	// L7 policies first, then L4 per: https://github.com/hashicorp/consul/blob/4e451f23584473a7eaf7f123145ca85e0a31783a/agent/xds/rbac.go#L664
+	for i, permission := range l7Perms {
+		policy := makeL7RBACPolicy(permission)
+		if policy != nil {
+			policies[fmt.Sprintf("%s-%d", baseL7PermissionKey, i)] = policy
+		}
+	}
+	for i, permission := range l4Perms {
+		policy := makeL4RBACPolicy(permission)
+		if policy != nil {
+			policies[l4PolicyLabel(i)] = policy
+		}
+	}
+
+	return policies
+}
+
+func makeL7RBACPolicy(p *pbproxystate.Permission) *envoy_rbac_v3.Policy {
+	if len(p.Principals) == 0 {
```

Comment:
I know this check is in the L4 equivalent too, but I'm confused about how this case is possible since we validate that sources on traffic permissions cannot be empty 🤔

Reply:
Given strange cases like this and how similar the L4 and L7 versions of this are, I wonder if we should try to share some of this logic to prevent future drift.

Reply:
I mainly saw it as a help manually generating proxystate for testing or advanced debugging, but would be fine with removing. I'd be happy to refactor this in the testing PR for a more unified flow. The separate configuration flows by port is what initially inclined to view the paths separately.